wallet: Properly rebroadcast unconfirmed transaction chains

achow101 opened the pull request:
Currently `ResendWalletTransactions` (used for normal rebroadcasts) will attempt to rebroadcast all of the transactions in the wallet in the order they are stored in `mapWallet`. This ends up being random as `mapWallet` is a `std::unordered_map`. However `ReacceptWalletTransactions` (used for adding to the mempool on loading) first sorts the txs by wallet insertion order, then submits them. The result is that `ResendWalletTranactions` will fail to rebroadcast child transactions if their txids happen to be lexicographically less than their parent's txid. This PR resolves this issue by combining `ReacceptWalletTransactions` and `ResendWalletTransactions` into a new `ResubmitWalletTransactions` so that the iteration code and basic checks are shared.

A test has also been added that checks that such transaction chains are rebroadcast correctly.

achow101 committed:
test: Test that an unconfirmed not-in-mempool chain is rebroadcast

The test checks that parent txs are broadcast before child txs.

The previous behavior is that the rebroadcasting would simply iterate mapWallet. As
mapWallet is a std::unsorted_map, the child can sometimes come before the parent and thus
be rebroadcast in the wrong order and fail the test.

glozow reviewed (commented):
Concept ACK

glozow commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -74,6 +79,41 @@ def run_test(self):
         node.setmocktime(now + 36 * 60 * 60 + 600)
         peer_second.wait_for_broadcast([txid])
 
+        self.log.info("Chain of unconfirmed not-in-mempool txs are rebroadcast")
```
Question: Is this test meant to fail without the first commit?

glozow replied:
It's passing for me on master :thinking:

achow101 replied:
Since master has `mapWallet` as a `std::unsorted_map`, this will now only fail randomly on it. It should fail half of the time, on average. I don't know how we can guarantee that the child will come before the parent in an unsorted map.

achow101 replied:
I've removed the "make a spend that will come before" and updated the comment to explain that.

glozow replied:
Hm, I see what you're saying about `std::unsorted_map` and it makes sense to me, but I've run the test >20 times and haven't gotten a failure yet...

achow101 replied:
Found a bug in the test. It should fail intermittently on master now

glozow replied:
Aha, fails intermittently on master for me now :) thanks!

kevkevinpal replied:
I'm seeing the test fail intermittently on commit [e40ddf3](https://github.com/bitcoin/bitcoin/pull/25768/commits/e40ddf36bed81bdf28d386eb961c9ed22b69e207) does that sound right or should the test only fail on the master branch

achow101 replied:
It should only fail on master. It should not fail on this PR. Make sure that you've compiled this PR before running the test.

kevkevinpal replied:
hmm I think I've compiled it correctly (still new to the repo so might be doing it incorrectly), this is what I did

```
git fetch origin pull/25768/head:PR25768
git checkout  PR25768
./autogen.sh
./configure
make clean
make -j "$(($(sysctl -n hw.physicalcpu)+1))"
./test/functional/wallet_resendwallettransactions.py
```

This is the error I see after I run the test, sometimes it passes

```
2022-08-18T20:38:25.914000Z TestFramework (INFO): Initializing test directory /var/folders/9g/cvx014yx4dq5lwl_x5zvn_j80000gn/T/bitcoin_func_test_43xe3o56
2022-08-18T20:38:28.879000Z TestFramework (INFO): Create a new transaction and wait until it's broadcast
2022-08-18T20:38:29.627000Z TestFramework (INFO): Create a block
2022-08-18T20:38:29.634000Z TestFramework (INFO): Bump time & check that transaction is rebroadcast
2022-08-18T20:38:29.905000Z TestFramework (INFO): Chain of unconfirmed not-in-mempool txs are rebroadcast
2022-08-18T20:38:31.037000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/kevkevin/DEVDIR/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_test()
  File "/Users/kevkevin/DEVDIR/bitcoin/./test/functional/wallet_resendwallettransactions.py", line 107, in run_test
    assert_raises_rpc_error(-5, "Transaction not in mempool", node.getmempoolentry, txid)
  File "/Users/kevkevin/DEVDIR/bitcoin/test/functional/test_framework/util.py", line 126, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
AssertionError: No exception raised
2022-08-18T20:38:31.089000Z TestFramework (INFO): Stopping nodes
2022-08-18T20:38:31.508000Z TestFramework (WARNING): Not cleaning up dir /var/folders/9g/cvx014yx4dq5lwl_x5zvn_j80000gn/T/bitcoin_func_test_43xe3o56
2022-08-18T20:38:31.508000Z TestFramework (ERROR): Test failed. Test logging available at /var/folders/9g/cvx014yx4dq5lwl_x5zvn_j80000gn/T/bitcoin_func_test_43xe3o56/test_framework.log
2022-08-18T20:38:31.508000Z TestFramework (ERROR):
2022-08-18T20:38:31.509000Z TestFramework (ERROR): Hint: Call /Users/kevkevin/DEVDIR/bitcoin/test/functional/combine_logs.py '/var/folders/9g/cvx014yx4dq5lwl_x5zvn_j80000gn/T/bitcoin_func_test_43xe3o56' to consolidate all logs
2022-08-18T20:38:31.509000Z TestFramework (ERROR):
2022-08-18T20:38:31.509000Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log.
2022-08-18T20:38:31.509000Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
2022-08-18T20:38:31.509000Z TestFramework (ERROR):
```

running on MacOs Monterey Version 12.3.1

achow101 replied:
I'm not seeing that error even when running on a loop.

MarcoFalke replied:
> git checkout  PR25768

Can you check that `git log -1` prints the correct hash, or alternatively just checkout the correct hash directly? Also, make sure that the working tree is clean.

```
git fetch origin e40ddf36bed81bdf28d386eb961c9ed22b69e207
git checkout e40ddf36bed81bdf28d386eb961c9ed22b69e207
git reset --hard HEAD # warning!! wipes unstaged changes in the tree if there still are some

kevkevinpal replied:
still getting the error and I am able to see I'm on the correct commit with `git log -1` and tried thee commands you posted aswell, I'll try and dig deeper to see what may be the issue

kevkevinpal replied:
not exactly sure what happened but after rebuilding master then rebuilding this PR's branch twice the test passes now sorry for any confusion

kevkevinpal commented on src/wallet/wallet.cpp:
```diff
@@ -1818,32 +1818,37 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
+        sorted_txs.insert(std::make_pair(wtx.nOrderPos, &wtx));
+    }
+    std::vector<CWalletTx*> out;
+    for (auto [pos, wtx] : sorted_txs) {
```
nit: not sure if we need `pos` here doesn't seem to be used

achow101 replied:
Done

kevkevinpal commented:
Concept ACK
was able to run the functional test on my local and it ran fine, still want to test on regtest but still new to the project so still learning

Riahiamirreza commented on src/wallet/wallet.cpp:
```diff
@@ -1818,32 +1818,37 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
+    // Sort pending transactions based on their initial wallet insertion order
+    std::map<int64_t, CWalletTx*> sorted_txs;
+    for (auto& [wtxid, wtx] : mapWallet) {
+        assert(wtx.GetHash() == wtxid);
```
Sorry, I don't understand the need for `assert(wtx.GetHash() == wtxid);`. In which case can this `assert` fails? Only when the data is corrupted?

sipa replied:
There is never a *need* for an assert. Assertions verify assumptions about the codebase, and should only fail if something you believe to be impossible in the codebase to be the case. It should *never* trigger, unless there is a bug in the code.

By writing this, the code author signals to readers of the code "here you can assume that wtxid is always equal to `wtx.GetHash()`". Further, this is verified during all tests, so we know it is also true for all scenarios any test case runs through.

glozow replied:
Wait, I see this is just moved and now new, but is this key wtxid or txid? Or does the "w" mean wallet instead of witness?

If it's wtxid, this assertion could fail when wtxid != txid since `CWalletTx::GetHash()` calls `tx->GetHash()`. But AFAICT it is a txid, so should be called txid instead? https://github.com/bitcoin/bitcoin/blob/02aefa169a9e6ed12c7bd8f3392adcd073d8d56b/src/wallet/wallet.cpp#L956-L971

achow101 replied:
The "w" refers to "wallet".

naumenkogs replied:
Is it good time to rename it to `wallet_txid` for clarification?
Without @glozow comment, i would just assume w stands for witness without any doubt...

hernanmarino replied:
The variable name also confused me , +1 to renaming it

pablomartin4btc replied:
I got the same confusion before reading the entire thread, thought it came from witness...

achow101 replied:
The variable names `wtxid` and `wtx` are used very frequently throughout the wallet code to refer to wallet transactions and their txids. If people find this confusing, then I think it would be more appropriate to have a scripted-diff PR rename all of those rather than renaming each individual instance that occurs in a PR.

achow101 replied:
Renamed to `txid`.

Riahiamirreza commented:
Would you explain what is the difference between `ResendWalletTransactions` and `ReacceptWalletTransactions`?  As you said `ResendWalletTransactions` is used for normal rebroadcasting and `ReacceptWalletTransactions` is used for rebroadcasting on loading. I'm not sure what do you mean by "loading", you mean something like reindex or reorg?

achow101 commented:
> I'm not sure what do you mean by "loading", you mean something like reindex or reorg?

"Loading" as in when the wallet is read from disk and loaded into memory.

Riahiamirreza commented:
So what is the difference between rebroadcasting when loading and in normal cases?

achow101 commented:
There are differences in which transactions will be considered for rebroadcasting. On load, all non-abandoned transactions will be rebroadcast. The periodic rebroadcast will only rebroadcast those that are more than 5 minutes older than the most recent block.

stickies-v commented on src/wallet/wallet.cpp:
```diff
@@ -1818,32 +1818,37 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
 
-        if (!wtx.IsCoinBase() && (nDepth == 0 && !wtx.isAbandoned())) {
-            mapSorted.insert(std::make_pair(wtx.nOrderPos, &wtx));
+        if (!wtx->IsCoinBase() && (nDepth == 0 && !wtx->isAbandoned())) {
```
When looking at `CWallet::GetTxDepthInMainChain`, if `nDepth == 0`, isn't that equivalent to `!wtx->IsConfirmed() && !wtx->IsConflicted()`, which in turn in combination with `!wtx->isAbandoned()` is equivalent to `!wtx->IsUnconfirmed()` - am I missing something? (I know this is moved code)
```suggestion
        if (!wtx->IsCoinBase() && !wtx->IsUnconfirmed()) {
```

achow101 replied:
Yes, those should be equivalent. Will do if I need to re-push.

achow101 replied:
Done

naumenkogs commented on src/wallet/wallet.cpp:
```diff
@@ -1818,32 +1818,37 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
     return result;
 }
 
+std::vector<CWalletTx*> CWallet::GetSortedTxs()
```
Should we specify what kind of sorting is applied?

stickies-v replied:
I was about to add the same suggestion 👍

achow101 replied:
The function has ended up being removed.

naumenkogs commented on src/wallet/wallet.cpp:
```diff
@@ -1914,15 +1919,14 @@ void CWallet::ResendWalletTransactions()
-        // Relay transactions
-        for (std::pair<const uint256, CWalletTx>& item : mapWallet) {
-            CWalletTx& wtx = item.second;
+        // Try to add wallet transactions to memory pool
```
"and relay"?

pablomartin4btc replied:
you meant the variable? the change is on `void CWallet::ResendWalletTransactions()`.

achow101 replied:
Done

1440000bytes commented:
> This PR resolves this issue by refactoring ReacceptWalletTransactions sorting code into a separate function, then using that function in both ReacceptWalletTransactions and ResendWalletTransactions so that both functions will sort the transactions in the same order before processing for rebroadcast.

Is it possible to shuffle transactions before `SubmitTxMemoryPoolAndRelay` rebroadcasting using [`std::shuffle`](https://en.cppreference.com/w/cpp/algorithm/random_shuffle). Since only wallet txs are rebroadcasted which is a separate privacy issue, we could at least hide the order in which they were inserted in wallet db.

rajarshimaitra commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -74,6 +79,41 @@ def run_test(self):
         peer_second.wait_for_broadcast([txid])
 
+        self.log.info("Chain of unconfirmed not-in-mempool txs are rebroadcast")
+        # Ideally we would add a transaction that would spend the existing transaction and also be
+        # in mapWallet before it. However mapWallet is a std::unsorted_map which means that it is
+        # hard for us to externally construct such a transaction. So we will ignore that fact and
+        # rely on the fact that because there are only two items in mapWallet, 50% of the time the
+        # child will be placed before the parent in mapWallet. The incorrect behavior that this
+        # test tries to catch should thus fail intermittently at a rate of 50%.
```
I am a bit confused on the intermittent test failure occurring here, maybe I am missing something.

Isn't sorting the order by wallet insertion index is what this PR fixes? So why would it still fail to reboadcast if `child.txid() < parent.txid()`? 

Also, does it make sense to have some unit testing to ensure this behavior of the `GetSorted()` function itself?

achow101 replied:
> Isn't sorting the order by wallet insertion index is what this PR fixes? So why would it still fail to reboadcast if `child.txid() < parent.txid()`?

The comment is talking about the code before this PR and how the test tries to catch it, not that this test should fail intermittently after the PR.

> Also, does it make sense to have some unit testing to ensure this behavior of the `GetSorted()` function itself?

There are already tests for the underlying `wtxOrdered` itself (or tests that end up testing that behavior, although perhaps not specifically).

rajarshimaitra replied:
So that means the code should not fail intermittently?? I will double check again that I am testing the correct binary, but I thoughts that's expected behavior from the comment..

achow101 replied:
It should only fail intermittently on master.

rajarshimaitra reviewed (commented):
Concept + tACK e40ddf36bed81bdf28d386eb961c9ed22b69e207

 - Verified that the tests are intermittently failing. 
 - Failure occurring at debug log assertion `ResendWalletTransactions: resubmit 2 unconfirmed transactions`
 
 Below are just one nit comment and a conceptual question.

achow101 commented:
> Is it possible to shuffle transactions before `SubmitTxMemoryPoolAndRelay` rebroadcasting using [`std::shuffle`](https://en.cppreference.com/w/cpp/algorithm/random_shuffle)? Since only wallet txs are rebroadcasted which is a separate privacy issue, we could at least hide the order in which they were inserted in wallet db.

No. Shuffling is the reason we have this issue in the first place (there is no order in an unsorted map so iterating it is basically the same as iterating a shuffled list). I don't think the insertion order is a privacy leak since it's the same order as when the transactions were originally broadcast which is public information to anyone that's connected to the network. Additionally, this is not a unilateral broadcast - it is up to the mempool relay code to determine when, how, and which transactions are actually rebroadcast. Only transactions that are not already in the mempool will be rebroadcast.

1440000bytes commented:
> No. Shuffling is the reason we have this issue in the first place (there is no order in an unsorted map so iterating it is basically the same as iterating a shuffled list). 

I was suggesting to shuffle the sorted list before rebroadcast. So we have all the transactions we need but do not relay in that order.

> I don't think the insertion order is a privacy leak since it's the same order as when the transactions were originally broadcast which is public information to anyone that's connected to the network.

In that case, feel free to ignore my suggestion.

1440000bytes commented:
Concept ACK

achow101 commented:
> I was suggesting to shuffle the sorted list before rebroadcast. So we have all the transactions we need but do not relay in that order.

Shuffling after sorting would just undo the sorting. Sorting is not used for filtering; it is used because we need to have the transactions in order. Specifically parent transactions must be rebroadcast before child transactions, and that specific order can be obtained through the insertion order.

While we could do something more complicated like actually building packages so that the actually order dependent things are grouped together, and then shuffling the packages, but I don't see the benefit in doing that. That might actually harm privacy because we would be broadcasting transactions that are related to each other at the same time, but only those transactions that are related to our wallet. Since a package that the wallet builds may exclude other transactions in the package that the mempool would know about, we could essentially be telling other nodes some of the scripts that our wallet is watching for as they could then look at what the rebroadcast package has that the rest of the package in the mempool does not.

achow101 committed:
wallet: Deduplicate Resend and ReacceptWalletTransactions

Both of these functions do almost the exact same thing. They can be
deduplicated so that their behavior matches except for the filtering
aspect. As this function will now always be called on wallet loading,
nNextResend will also always be initialized, so
wallet_resendwallettransactions.py is updated to account for that.

This also resolves a bug where ResendWalletTransactions would fail to
rebroadcast txs in insertion order thereby potentially rebroadcasting a
child transaction before its parent and causing the child to not
actually get rebroadcast.

Also names the combined function to ResubmitWalletTransactions as the
function just submits the transactions to the mempool rather than doing
any sending by itself.

achow101 commented:
Based on some discussions during the PR review club today, I've made several changes to this PR. The first is that `ResendWalletTransaction` and `ReacceptWalletTransactions` are now combined into a single function `ResubmitWalletTranasctions`. There were some subtle differences between the two functions that are being preserved by using two optional parameters. The first difference is that `ReacceptWalletTransactions` would not actually ask the mempool to relay the transactions (there is a parameter for relay in `SubmitTxMemoryPoolAndRelay` that was different in both). There is a parameter `relay` with a default of `true` to handle that. The second is that `ResendWalletTransactions` would look chain status (is ibd, reindexing, etc.) and transaction time in order to determine whether to try submitting a transaction, whereas `ReacceptWalletTransactions` would not. To match the `ReaccpetWalletTransactions` behavior, a `force` option with a default of `false` is added that will bypass those checks.

Additionally, `nNextResend` was initialized the first time `ResendWalletTransactions` was run. As the new `ResubmitWalletTransactions` is always run on start, I've removed from the test the need to force it to be run as well as the handling around not actually trying to resend during that first run (we want to do the reaccept behavior in that first run). I've also removed `GetSortedTxs` and have instead taken the suggestion of filtering then sorting the transactions. This is done by inserting the `CWalletTx`s into a `std::set` with a custom comparator that uses `CWalletTx.nOrderPos` to determine the sort order.

Lastly I've updated the test to reliably fail on master. I found that `listreceivedbyaddress` and related RPCs will output txids in the order that they are found in `mapWallet`. The new test will use that to check if the child is positioned before the parent, and if it is not, it will remake the child until it does.

achow101 commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -27,7 +29,10 @@ def run_test(self):
+        txid = node.send(outputs=[{node.getnewaddress(): 1}], options={"inputs": [parent_utxo]})["txid"]
 
         # Wallet rebroadcast is first scheduled 1 sec after startup (see
         # nNextResend in ResendWalletTransactions()). Tell scheduler to call
```
The comment is removed now that this behavior is no longer needed.

achow101 commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -27,7 +29,10 @@ def run_test(self):
         self.log.info("Create a new transaction and wait until it's broadcast")
-        txid = node.sendtoaddress(node.getnewaddress(), 1)
+        unspent = node.listunspent()
+        parent_utxo = unspent[0]
+        indep_utxo = unspent[1]
```
I've reimplemented the test so that it always fails on master.

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25768/commits/fee0dea1b1e32832dc5efca3621cfc8ba7c6c8f9

hernanmarino reviewed (approved):
ACK fee0dea

Now the test consistently fails on master, as it should.

naumenkogs commented on src/wallet/wallet.cpp:
```diff
@@ -1886,43 +1858,57 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+    // even if forcing.
+    if (!fBroadcastTransactions) return;
+
     // During reindex, importing and IBD, old wallet transactions become
```
I'd say this comment is now slightly outdated. My first thought was "so all those calls with force=true will spam now?". Then I realized they all have `relay=false`.

I think extending this comment would improve readability. Might as well document what force does exactly (in broader terms is fine).

achow101 replied:
I've expanded the comment and also added `relay` to the check. This way we guarantee that we won't spam if we want to relay, regardless of `force`.

naumenkogs commented on src/wallet/wallet.cpp:
```diff
@@ -1886,43 +1858,57 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+    if (!force && GetTime() < nNextResend) return;
     // resend 12-36 hours from now, ~1 day on average.
     nNextResend = GetTime() + (12 * 60 * 60) + GetRand(24 * 60 * 60);
-    if (fFirst) return;
```
It took me some time to realize why it's safe. I think it's worth expanding `On startup, we will do this for all unconfirmed // transactions but will not ask the mempool to relay them.` saying it protects privacy as a side effect?

achow101 replied:
Done.

naumenkogs commented on src/wallet/wallet.cpp:
```diff
@@ -1886,43 +1858,57 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+            // For periodic rebroadcast (force == false),
+            // attempt to rebroadcast all txes more than 5 minutes older than
+            // the last block.
+            if (!force && wtx.nTimeReceived > m_best_block_time - 5 * 60) continue;
```
Slightly off-topic, is this code prone to privacy leaks? Imagine a fork appearing: one (empty) block with a timestamp in the past. Would many own transactions get rebroadcast? If yes, i assume other nodes won't forward them because they had them already, and it becomes apparent who is the source?

achow101 replied:
I don't think so. Your own transactions would be submitted to your node's mempool first, and only if it does not know about those transactions will it try to broadcast them.

naumenkogs replied:
Right, thank you.

rajarshimaitra reviewed (approved):
tACK fee0dea1b1e32832dc5efca3621cfc8ba7c6c8f9

Verified the test is failing reliably in master.. 

Just two more nit comments (questions)..

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1886,43 +1858,57 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+        for (auto wtx : to_submit) {
             std::string unused_err_string;
-            if (SubmitTxMemoryPoolAndRelay(wtx, unused_err_string, true)) ++submitted_tx_count;
+            if (SubmitTxMemoryPoolAndRelay(*wtx, unused_err_string, relay)) ++submitted_tx_count;
```
No. We have had a new style in effect for the past several years. It only applies to new code, so a bunch of the old code written under the old style is still around.

achow101 commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -66,14 +66,61 @@ def run_test(self):
             # Tell scheduler to call MaybeResendWalletTxn now.
             node.mockscheduler(1)
         # Give some time for trickle to occur
         node.setmocktime(now + 36 * 60 * 60 + 600)
```
`mockscheduler` does not change the mocked time.

stickies-v commented on src/wallet/transaction.h:
```diff
@@ -305,6 +305,13 @@ class CWalletTx
     void operator=(CWalletTx const &x) = delete;
 };
+
+struct WalletTxOrderComparator {
```
What do you think about a lambda function instead? Quite a bit less code, and imo about equally readable? As long as we're not using this comparator anywhere else the lambda would have my preference, but I'm okay with the current implementation too.

<details>
<summary> git diff </summary>

```diff
diff --git a/src/wallet/transaction.h b/src/wallet/transaction.h
index 27983e356..c36cbeb23 100644
--- a/src/wallet/transaction.h
+++ b/src/wallet/transaction.h
@@ -306,12 +306,6 @@ public:
     void operator=(CWalletTx const &x) = delete;
 };
 
-struct WalletTxOrderComparator {
-    bool operator()(const CWalletTx* a, const CWalletTx* b) const
-    {
-        return a->nOrderPos < b->nOrderPos;
-    }
-};
 } // namespace wallet
 
 #endif // BITCOIN_WALLET_TRANSACTION_H
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 69957ed9e..ad20f332a 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1906,8 +1906,9 @@ void CWallet::ResubmitWalletTransactions(bool relay, bool force)
         LOCK(cs_wallet);
 
         // First filter for the transactions we want to rebroadcast.
-        // We use a set with WalletTxOrderComparator so that rebroadcasting occurs in insertion order
-        std::set<CWalletTx*, WalletTxOrderComparator> to_submit;
+        // We use a set ordered by nOrderPos so that rebroadcasting occurs in insertion order
+        auto cmp = [](const CWalletTx* a, const CWalletTx* b) { return a->nOrderPos < b->nOrderPos; };
+        std::set<CWalletTx*, decltype(cmp)> to_submit(cmp);
         for (auto& [txid, wtx] : mapWallet) {
             // Only rebroadcast unconfirmed txs
             if (!wtx.isUnconfirmed()) continue;
```

</details>

achow101 replied:
I think the comparator might actually be useful in other places too.

stickies-v reviewed (approved):
utACK 71b7709fa075dc4c9b55bb07304687724f67f0cb

The refactoring into `ResubmitWalletTransactions()` makes a lot of sense and improves understanding the codebase imo.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1886,43 +1858,70 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+// that depends on the `relay` option. Periodic rebroadcast uses the pattern
+// relay=true force=false (also the default values), while loading into the mempool
+// (on start, or after import) uses relay=false force=true.
+void CWallet::ResubmitWalletTransactions(bool relay, bool force)
```
I agree that `force` implies that more will happen than actually does, but I couldn't think of anything better to call it. I don't think `force_ready` is any clearer.

achow101 commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -27,7 +29,10 @@ def run_test(self):
-        txid = node.sendtoaddress(node.getnewaddress(), 1)
+        unspent = node.listunspent()
+        parent_utxo = unspent[0]
+        indep_utxo = unspent[1]
```
Done

achow101 commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -74,6 +79,41 @@ def run_test(self):
+        node.syncwithvalidationinterfacequeue()
+
+        # Evict these txs from the mempool
+        evict_time = block_time + 60 * 60 * 336 + 5
```
Done

achow101 commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -74,6 +79,41 @@ def run_test(self):
+
+        # Rebroadcast and check that parent and child are both in the mempool
+        with node.assert_debug_log(['ResendWalletTransactions: resubmit 2 unconfirmed transactions']):
+            node.setmocktime(evict_time + 36 * 60 * 60)
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1886,43 +1858,70 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+
+            // For periodic rebroadcast (force == false),
+            // attempt to rebroadcast all txes more than 5 minutes older than
+            // the last block.
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1886,43 +1858,70 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+// We do this on a random timer to slightly obfuscate which transactions
+// come from our wallet.
 //
 // Ideally, we'd only resend transactions that we think should have been
```
Done

naumenkogs commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -68,6 +73,53 @@ def run_test(self):
         peer_second.wait_for_broadcast([txid])
 
+        self.log.info("Chain of unconfirmed not-in-mempool txs are rebroadcast")
+        # This tests that the node broadcasts the parent transaction before the parent transaction.
```
`the parent transaction before the parent transaction`?

achow101 replied:
Oops, fixed.

naumenkogs commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -68,6 +73,53 @@ def run_test(self):
 
+        self.log.info("Chain of unconfirmed not-in-mempool txs are rebroadcast")
+        # This tests that the node broadcasts the parent transaction before the child transaction.
+        # The previous behavior is that the rebroadcasting would simply iterate mapWallet. As
```
nit: I think `the previous behavior` stuff belongs to the commit message, not to the file content.

achow101 replied:
Done

LarryRuane reviewed (approved):
ACK 748d5c75d60505a3d84d36782f5cf4949224e103

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1936,7 +1935,7 @@ void CWallet::ResendWalletTransactions()
 {
     for (const std::shared_ptr<CWallet>& pwallet : GetWallets(context)) {
-        pwallet->ResendWalletTransactions();
+        pwallet->ResubmitWalletTransactions();
```
Removed the default arguments.

1440000bytes reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/25768/commits/3405f3eed5cf841b23a569b64a376c2e5b5026cd

naumenkogs commented:
utACK 3405f3eed5cf841b23a569b64a376c2e5b5026cd

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1925,43 +1897,69 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
     // During reindex, importing and IBD, old wallet transactions become
     // unconfirmed. Don't resend them as that would spam other nodes.
-    if (!chain().isReadyToBroadcast()) return;
+    // We only allow forcing mempool submission when not relaying to avoid this spam.
+    if (!force && relay && !chain().isReadyToBroadcast()) return;
 
     // Do this infrequently and randomly to avoid giving away
     // that these are our transactions.
-    if (GetTime() < nNextResend || !fBroadcastTransactions) return;
-    bool fFirst = (nNextResend == 0);
+    if (!force && GetTime() < nNextResend) return;
```
At this point of the review, this is just a minor nit but code would be a bit clearer in this way:
```c++
if (!force) {
    // During reindex, importing and IBD, old wallet transactions become
    // unconfirmed. Don't resend them as that would spam other nodes.
    // We only allow forcing mempool submission when not relaying to avoid this spam.
    if (relay && !chain().isReadyToBroadcast()) return;

    // Do this infrequently and randomly to avoid giving away
    // that these are our transactions.
    if (GetTime() < nNextResend) return;
}
```

stickies-v replied:
I think both these checks would be more appropriate in `MaybeResendWalletTxs()` since they're more scheduling-like in nature. It would also address furszy's comment them since we don't need to check for `force` anymore (it's always True there). If you do this, you probably also want to duplicate `if (!fBroadcastTransactions)` there since it's a very cheap check.

Unfortunately, we still need the `force` parameter when iterating over the transactions, otherwise `ResubmitWalletTransactions()`'s signature could have been simplified.

Tetita replied:
:)  Bien dicen que peca más el que lo piensay más NO el que lo hace.
xD

furszy reviewed (approved):
Late code review ACK 3405f3ee

The only orthogonal question that arises from my side is why are we trying to resubmit txes after every import RPC command. Can anything change at the, already created, wallet txes level after an import?

MarcoFalke commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -66,14 +65,57 @@ def run_test(self):
+        evict_time = block_time + 60 * 60 * DEFAULT_MEMPOOL_EXPIRY_HOURS + 5
+        node.setmocktime(evict_time)
+        indep_send = node.send(outputs=[{node.getnewaddress(): 1}], options={"inputs": [indep_utxo]})
+        node.syncwithvalidationinterfacequeue()
```
Can you explain why this is needed? Generally this is only needed when forcing the wallet to receive all in-flight mempool notifications. However, there is no wallet call that follows. Also, there is no `m_best_block_time` update.

achow101 replied:
> Also, there is no `m_best_block_time` update.

Why not? IIRC this was done in order to update `m_best_block_time` so that the transaction would be properly evicted later.

MarcoFalke replied:
The transaction is evicted from the mempool, not the wallet. (Only the wallet has an `m_best_block_time`)

Also, there is no new best block, so `m_best_block_time` won't be updated either way.

achow101 replied:
Oh, wrong line. I think this `syncwithvalidationinterfacequeue` is probably unneeded.

MarcoFalke commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -66,14 +65,57 @@ def run_test(self):
+        block = create_block(int(node.getbestblockhash(), 16), create_coinbase(node.getblockcount() + 1), block_time)
+        block.solve()
+        node.submitblock(block.serialize().hex())
+        node.syncwithvalidationinterfacequeue()
```
Can you explain why this is needed? Maybe duplicate the comment "Set correct m_best_block_time" from above?

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1925,43 +1897,69 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+// The `force` option results in all unconfirmed transactions being submitted to
+// the mempool. This does not necessarily result in those transactions being relayed,
+// that depends on the `relay` option. Periodic rebroadcast uses the pattern
+// relay=true force=false (also the default values), while loading into the mempool
```
```suggestion
// relay=true force=false, while loading into the mempool
```

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1925,43 +1897,69 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
+            // Only rebroadcast unconfirmed txs
+            if (!wtx.isUnconfirmed()) continue;
+
+            // attempt to rebroadcast all txes more than 5 minutes older than
```
```suggestion
            // Attempt to rebroadcast all txes more than 5 minutes older than
```

Any reason that this `A` changed into `a`?

maflcko reviewed (commented):
left some minor nits, which can be addressed the next time this code is touched

stickies-v reviewed (commented):
ACK 3405f3eed5cf841b23a569b64a376c2e5b5026cd

MarcoFalke commented:
If a user calls `importaddress` every 6 hours, this will block any p2p relay from ever happening?

stickies-v commented:
> If a user calls importaddress every 6 hours, this will block any p2p relay from ever happening?

I addressed that in #26205
